src/ops_complex.rs

```diff
@@ -605,15 +600,25 @@ pub fn generate_permutation_matrix(
         32,
         stream,
     )?;
+    let (unique_variable_indexes, temp_storage) = temp_storage.split_at_mut(num_bytes);
+    let unique_variable_indexes = unsafe { unique_variable_indexes.transmute_mut() };
+    let (run_lengths, temp_storage) = temp_storage.split_at_mut(num_bytes);
+    let run_lengths = unsafe { run_lengths.transmute_mut() };
+    let (num_runs_out, encode_temp_storage) = temp_storage.split_at_mut(cell_size);
+    let num_runs_out = unsafe { &mut num_runs_out.transmute_mut()[0] };
     set_to_zero(run_lengths, stream)?;
     encode(
-        temp_storage,
+        encode_temp_storage,
         sorted_variable_indexes,
         unique_variable_indexes,
         run_lengths,
         num_runs_out,
         stream,
     )?;
+    let (run_offsets, run_indexes) = temp_storage.split_at_mut(num_cells * 4);
+    let run_offsets = unsafe { run_offsets.transmute_mut() };
+    let run_indexes = unsafe { run_indexes.transmute_mut() };
```

Comment:
After populating `run_indexes` below, why do we need to keep `run_lengths` around? If the scan is inclusive, I think threads in `mark_ends_of_runs` and `generate_permutation_matrix_raw` could locally reconstruct
```cuda
my_run_length = run_index == 0 ? run_lengths[0] : run_lengths[run_index] - run_lengths[run_index  - 1];
```

(But I don't think dropping `run_lengths` helps reduce the memory high water mark overall :( )

Reply:
i implemented this change, now the memory requirement is `(4 * num_cells + 1) * sizeof::<u32>()`